drivers: can: mcp2515: dts fixes

karstenkoenig opened the pull request:
Adding required fields to the devicetree overlay of the MCP2515 as
this is often used as a reference and use these fields instead of the
KConfig entries. Also turned it into a shield to align this with the newer style.

karstenkoenig committed:
drivers: can: mcp2515: fix devicetree bindings

Adding required fields to the devicetree overlay of the CAN sample as
this is often used as a reference. Also use these fields instead of the
KConfig entries.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig committed:
shields: mcp2515: Add simple mcp2515 breakout board

The mcp2515 is often provided as a simple breakout board including CAN
tranceiver, so this config and overlay provide a good base for that.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented:
This should address https://github.com/zephyrproject-rtos/zephyr/issues/17926

karstenkoenig commented:
@nwsetec 
Can you have a look maybe as well? on current master the driver shouldn't build without changes along what's being done here as overlay didn't satisfy some requirements. Or did you create your own board device tree and added it there plus those required fields such as 'prop-seg'?

nwsetec commented:
@karstenkoenig sure, my branch is a fair way behind master, I'll try to rebase and try this out tonight.  So far I've got it working with an overlay.

alexanderwachter commented:
@nwsetec you can also have a look at #18081. This PR is the new sample code.

nwsetec commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -20,34 +20,6 @@ config CAN_MCP2515_OSC_FREQ
-config CAN_PROP_SEG
-	int "Prop_Seg"
-	default 2
-	range 1 8
```
@karstenkoenig  I'm not sure if this is a good idea or not but as a user I have appreciated the information the range property Kconfig gives me (and stops me from doing bad things), is there any equivalent to check the device tree property settings?  The first solution that come to mind is build asserts.  There are runtime asserts in mcp2515_configure() but we could add build asserts for these CAN properties and possibly remove some runtime asserts.  Thoughts?

karstenkoenig replied:
Good idea, I converted the runtime asserts to buildtime asserts!

nwsetec commented on dts/bindings/can/microchip,mcp2515.yaml:
```diff
@@ -18,3 +18,6 @@ properties:
       type: compound
       category: required
+    reg:
+      type: array
```
Sorry for my ignorance, what does `reg:` and `type: array` mean hear for the binding?

galak replied:
`reg:` is the property name expected in the dts.  `array` says its an integer list of items.

karstenkoenig replied:
It doesn't really mean anything for the mcp2515, usually it would be a memory address or the i2c bus address, in this case it's from the spi-device.yaml. The 'reg' propery was already in the overlay dts file, but now I had to add it to yaml, otherwise the dts parser assumed the property was modified from required to optional. I couldn't figure out why, but this small change fixed that.

nwsetec replied:
Ok cheers, so currently it's a requirement that is fulfilled with the <0> in the overlay.

karstenkoenig replied:
@galak dealt with the matter here in a cleaner way
https://github.com/zephyrproject-rtos/zephyr/pull/18191

nwsetec commented:
> @nwsetec you can also have a look at #18081. This PR is the new sample code.

@alexanderwachter I'll try to have a look tomorrow

nwsetec commented on boards/shields/mcp2515/mcp2515.conf:
```diff
@@ -5,9 +5,5 @@ CONFIG_CAN_INIT_PRIORITY=80
 CONFIG_LOG=y
-CONFIG_CAN_LOG_LEVEL=4
\ No newline at end of file
+CONFIG_CAN_LOG_LEVEL=4
```
@karstenkoenig I think CONFIG_LOG=y and CONFIG_CAN_LOG_LEVEL=4 should be removed for a shield/breakout board?  Also I think it's missing CONFIG_GPIO=y and CONFIG_SPI=y?

karstenkoenig replied:
Removed CONFIG_LOG stuff

CONFIG_SPI is there, CONFIG_GPIO I don't know, I'd assume that is pulled in elsewhere to be honest. Is it common to add extra?

karstenkoenig committed:
drivers: can: mcp2515: build instead of runtime assert

Converted some runtime asserts to build time asserts to fail as early as
possible when an invalid configuration is present.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

nwsetec commented on drivers/can/can_mcp2515.c:
```diff
@@ -606,6 +594,25 @@ static struct mcp2515_data mcp2515_data_1 = {
+	"1 <= PROP <= 8");
+BUILD_ASSERT_MSG((DT_INST_0_MICROCHIP_MCP2515_PHASE_SEG1 >= 1) &&
+	(DT_INST_0_MICROCHIP_MCP2515_PHASE_SEG1 <= 8),
+	"1 <= BS1 <= 8");
```
Looking good!  For consistency should BS1, BS2 and PROP be PHASE_SEG1, PHASE_SEG2 and PROP_SEG?

karstenkoenig replied:
Phew to be honest no idea why I named them BS1 and BS2, should be PS1/ps1 and PS2/ps2, weird. I think I mainly wanted short names because the 80 char character limit is tough enough as is with all the comparisons in the checks.

nwsetec commented on drivers/can/can_mcp2515.c:
```diff
@@ -618,10 +625,10 @@ static const struct mcp2515_config mcp2515_config_1 = {
-	.tq_bs2 = CONFIG_CAN_PHASE_SEG2,
+	.tq_sjw = DT_INST_0_MICROCHIP_MCP2515_SJW,
+	.tq_prop = DT_INST_0_MICROCHIP_MCP2515_PROP_SEG,
+	.tq_bs1 = DT_INST_0_MICROCHIP_MCP2515_PHASE_SEG1,
```
Similarly here for the bs1, bs2 and prop.  Less mental translation between DT and driver state.

nwsetec commented on drivers/can/can_mcp2515.c:
```diff
@@ -606,6 +594,25 @@ static struct mcp2515_data mcp2515_data_1 = {
+	DT_INST_0_MICROCHIP_MCP2515_PHASE_SEG2,
+	"PROP + BS1 >= BS2");
+BUILD_ASSERT_MSG(DT_INST_0_MICROCHIP_MCP2515_PHASE_SEG2 >
+	DT_INST_0_MICROCHIP_MCP2515_SJW, "BS2 > SJW");
```
I'm checking the datasheet and I think the last two build assert requirements:
`PROP + BS1 >= BS2`
`BS2 > SJW`
are guidelines not strict requirements.  Better as warnings in mcp2515_init()?  Sorry I didn't see that before.

karstenkoenig replied:
They are in the datasheet at 5.5 as a requirement, are you sure it's not fixed?

karstenkoenig commented:
I am dropping this PR, the most important fix is in this https://github.com/zephyrproject-rtos/zephyr/pull/18081 as well and that PR is in pretty good shape. Plus apparently there is a proper arduino shield with the mcp2515 which is probably a better base than this overlay file

alexanderwachter commented:
drivers: can: mcp2515: build instead of runtime assert Is still usefull.

karstenkoenig commented:
> ‚drivers: can: mcp2515: build instead of runtime assert‘ Is still usefull.

Yeah I agree, I noticed DTS also supports ranges though, so I wanted to align it with this
https://github.com/zephyrproject-rtos/zephyr/pull/18271
And it does make sense to aim for a proper arduino shield with the pinout, so there will be a somewhat clear chipselect and IRQ pin and it's straight forward to connect it to various boards :-)
